pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,151 @@
+// As the create instance operation is in parallel, it is possible that an IP range was reserved by the service instance but is not updated in the cloud instance. The reserved list stores such IP ranges
+type IPAllocator struct {
+	reserved map[string]bool
+}
+
+// NewIPAllocator is the constructor to initialize the IPAllocator object
+// Argument reserved map[string]bool is a set of currently reserved IPs which have not been updated by the cloud provider
+func NewIPAllocator(reserved map[string]bool) *IPAllocator {
+	return &IPAllocator{
+		reserved: reserved,
+	}
+}
+
+// Reserve function adds a particular IP block in the reserve pool
+// Argument ipBlock string is an IPV4 range which needs to be reserved
+func (ipAllocator *IPAllocator) Reserve(ipV4Range string) {
+	ipAllocator.reserved[ipV4Range] = true
+}
+
+// Unreserve function frees up a reserved IP
+// Argument ipBlock string is an IPV4 range which needs to be unreserved
+func (ipAllocator *IPAllocator) Unreserve(ipV4Range string) {
+	delete(ipAllocator.reserved, ipV4Range)
+}
+
+// ValidateCIDR function validates whether a particular cidr is a valid IP range
+// Argument cidr string is is a CIDR range that needs to be validated
+func (ipAllocator *IPAllocator) ValidateCIDR(cidr string) error {
+	_, _, err := net.ParseCIDR(cidr)
+	return err
+}
+
+// GetUnreservedIPBlock returns an unreserved /29 IP block. It accepts the requested
+// CIDR and list of reserved IPs as arguments and returns the /29 IP available in that CIDR
+// Reserved IPV4 ranges can be available in the cloud instances or not as it is yet to be written.
+// All the Reserved IP ranges present in the cloud provider are passed to this function
+// All the Reserved IP ranges not present in the cloud provider are extracted from the reserved list in the IPAllocator
+// Finally a final reserved list is created by mergin these two lists
+// Potential error cases: 1) No /29 Block in the CIDR is unreserved
+// 2) Parsing the CIDR resulted in an error
+func (ipAllocator *IPAllocator) GetUnreservedIPBlock(cidr string, reservedIPRanges map[string]bool) (string, error) {
+	for reservedIPRange := range ipAllocator.reserved {
+		reservedIPRanges[reservedIPRange] = true
+	}
+
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", fmt.Errorf("Unable to parse CIDR %s", cidr)
+	}
+	buffer := bytes.NewBufferString("")
+	for cidrIPBlock := cloneIP(ip.Mask(ipnet.Mask)); ipnet.Contains(cidrIPBlock) && err == nil; err = incrementIP(cidrIPBlock, incrementStep29IPBlock) {
+		overLap := false
+		buffer.WriteString(cidrIPBlock.String())
+		buffer.WriteString("/29")
+		ipRange := buffer.String()
+		for reservedIPRange := range reservedIPRanges {
+			// Find if the current IP range in the CIDR overlaps with any of the reserved IP ranges. If not, this can be returned
+			overLap = isOverlap(ipRange, reservedIPRange)
+			if overLap {
+				break
+			}
+		}
+		if !overLap {
+			return ipRange, nil
+		}
+		buffer.Reset()
```

Comment:
same concern as last time. I don't think you need to keep writing to a buffer and resetting it here. When you find the correct IP range you can create it on the spot.